posts/loggingPython.md

```diff
@@ -0,0 +1,325 @@
+
+A generic log message provides just about as little information as no log message. Imagine if you had to go through your logs and you saw `purchase completed`. This doesn’t help you answer any questions. *When was the purchase completed? Who completed it? What did they buy?*
+
+This can all be done by formatting your logs:
+```python
+logFormatter = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
+logging.basicConfig(format=logFormatter)
+```
+
+### Structured Data
+
+The _biggest_ issue we’ve seen is the pollution of log messages. People will put objects or variables into the log. Though it’s helpful while debugging your code, it’s impossible to decipher what you were thinking when looking at the stream in the future. The best way to deal with this is to push all your metadata to the `extra` object. Using this, you’ll be able to decipher your messages from the stream, while still grab the full context to dig deeper into your user from any messages.
+
+Some of this data, such as `time` and `levelname` can be captured automatically, but if you can (and should) push `extra` context to your logs.
+
+```python
+logger = logging.getLogger(__name__)
+logger.info('purchase completed', extra={‘user': 'Sid Panjwani'})
+```
+
+If you’re wondering why you do not see your logs, remember to change the `level`. By default, it’s set to `warning`.
+
+*Yep, it’s that’s easy…*
+
+### Tracking Exceptions through Logs
+
+Some companies have made it their core competency helping you capture and react to exceptions. If you’re just a hobbyist, it’s pretty easy to capture your errors yourself and write them to your logs.
+
+```python
+def captureException():
+    # this should do something
+
+try:
+    1/0
+except:
+    captureException()
+```
+
+When capturing an exception, it’s essential to add context. Similar to when you’re debugging, you want to know who the user is and what they were doing when the exception struck.
+
+```python
+import sys
+
+# remember to set up your logger
+
+def captureException():
+    logger.warning(sys.exc_info())
+```
+
+![](./images/loggingPython/exception1.png)
+
+This gives us a `traceback` object that we can use to see the stack from our exception. We didn’t even need to pass anything into our `captureException` method.
+
+Let’s figure out what the `traceback` object gives us and clean up this information a little.
+
+```python
+import sys
+import traceback
+
+def captureException():
+    r = list(sys.exc_info())
+
+    e = dict()
+    e["name"] = r[1]
+```
+
+![](./images/loggingPython/exception2.png)
+
+Now, this can be cleaned up using some string manipulation in Python.
+
+## Storing & Accessing These Logs
+
+Now that you've learned to write these (beautiful) logs, you've got to learn what to do with them. I mean it would be a waste to write them to the console and leave them to be deleted when the program closes.
```

Comment:
I'd recommend you start off with a better explanation of what is happening by default, for example: "By default, logs will be written to the standard output device (which may be your terminal window), but Python's logging module offers a rich set of options for customizing the output handling to suit your needs." Note that logging to standard output is a standard encouraged by The Twelve-Factor App practice. Platforms like Heroku, Amazon Elastic Beanstalk, and Docker build on this standard by capturing standard output and redirecting it to other logging facilities at a platform level.